src/thor/costmatrix.cc

```diff
@@ -260,249 +278,390 @@ void CostMatrix::SourceToTarget(Api& request,
-                              pred.path_distance() + directededge->length(),
-                              (pred.not_thru_pruning() || !directededge->not_thru()),
-                              (pred.closure_pruning() || !costing_->IsClosed(directededge, tile)),
-                              static_cast<bool>(flow_sources & kDefaultFlowMask),
-                              costing_->TurnType(pred.opp_local_idx(), nodeinfo, directededge),
-                              restriction_idx);
-      adj.add(idx);
-
-      // setting this edge as reached
-      if (expansion_callback_) {
-        expansion_callback_(graphreader, edgeid, "costmatrix", "r", pred.cost().secs,
-                            pred.path_distance(), pred.cost().cost);
+      // setup for expansion at this level
+      hierarchy_limits[node.level()].up_transition_count += trans->up();
+      const auto* trans_node = trans_tile->node(trans->endnode());
+      EdgeMetadata trans_meta =
+          EdgeMetadata::make(trans->endnode(), trans_node, trans_tile, edgestatus);
+      uint32_t trans_shortcuts = 0;
+      // expand the edges from this node at this level
+      for (uint32_t i = 0; i < trans_node->edge_count(); ++i, ++trans_meta) {
+        disable_uturn = ExpandInner<expansion_direction>(graphreader, index, pred, opp_pred_edge,
+                                                         trans_node, pred_idx, trans_meta,
+                                                         trans_shortcuts, trans_tile, offset_time) ||
+                        disable_uturn;
       }
     }
+  }
 
-    // Handle transitions - expand from the end node of the transition
-    if (!from_transition && nodeinfo->transition_count() > 0) {
-      const NodeTransition* trans = tile->transition(nodeinfo->transition_index());
-      for (uint32_t i = 0; i < nodeinfo->transition_count(); ++i, ++trans) {
-        if (trans->up()) {
-          hierarchy_limits[node.level()].up_transition_count++;
-        } else if (hierarchy_limits[trans->endnode().level()].StopExpanding()) {
-          continue;
-        }
+  // Now, after having looked at all the edges, including edges on other levels,
+  // we can say if this is a deadend or not, and if so, evaluate the uturn-edge (if it exists)
+  if (!disable_uturn && uturn_meta) {
+    // If we found no suitable edge to add, it means we're at a deadend
+    // so lets go back and re-evaluate a potential u-turn
+    pred.set_deadend(true);
 
-        // Expand from end node of this transition.
-        GraphId node = trans->endnode();
-        graph_tile_ptr endtile = graphreader.GetGraphTile(node);
-        if (endtile != nullptr) {
-          expand(endtile, node, endtile->node(node), pred, pred_idx, true, ti);
-        }
-      }
-    }
-  };
+    // TODO(nils): what if there is a shortcut that supersedes our u-turn? can that even be?
+    // We then need to decide if we should expand the shortcut or the non-shortcut edge...
```

Comment:
this is a good question. yes we can make a uturn where a shortcut begins/ends. however, isnt it always safe to expand on the non shortcut? the worst scenario is that we were slightly less efficient in finding a solution because there as a shortcut that we didnt use. maybe thats all you're really saying here: "to be fully optimial we shoudl favor a shortcut uturn (when using shortcuts) as opposed to a superseded edge"